src/app/recipes/recipe.service.ts

```diff
@@ -24,11 +26,31 @@ export class RecipeService {
   ];
 
   getRecipeById(id: number) {
-    return this.recipes[id];
+    return this.recipes.filter((recipe: Recipe) => { return recipe.id === id; })[0];
   }
 
   getRecipes() {
     return this.recipes.slice();
   }
 
+  addRecipe(recipe: Recipe) {
+    recipe.id = this.recipes.length;
```

Comment:
Is it reliable id generation? What if 2nd recipe will be removed and new one added?

Reply:
This implementation works. However, I understand that frontend should not generate an id.

Reply:
It's not reliable implementation. And the problem is not in front-end assumption that length will not be changed is wrong since you can remove recipes. 
Let's imagine you have 3 recipes. The 2nd is removed. After that you add one more. Which id will be applied for new one ? '3' ?  as result you have duplicates and code like: return this.recipes.filter((recipe: Recipe) => { return recipe.id === id; })[0]; is required